File: api/app/controllers/api/v1/dashboard_controller.rb
class Api::V1::DashboardController < Api::V1::BaseController
  def show
    month = params[:month].presence || Time.zone.today.strftime("%Y-%m")
    month_start = Time.zone.parse("#{month}-01").beginning_of_month
    month_end   = month_start.end_of_month.end_of_day

    # 暫定: device_id は param 優先、なければ最初のデバイス
    device_id = params[:device_id].presence || Device.first&.id
    return render json: { error: "device not found" }, status: 404 unless device_id

    rel = BowlSnapshot.where(device_id: device_id, recorded_at: month_start..month_end)

    #  残りのごはんを取得
    last_snapshot = BowlSnapshot.where(device_id: device_id).order(recorded_at: :desc).first
    bowl_remaining = last_snapshot&.weight_g.to_f

    # === 今月の日別合計（0 埋め含む） ===
    # DATE(recorded_at) ごとに SUM(weight_g)
    raw_daily = rel
                .select("DATE(recorded_at) AS d, SUM(weight_g) AS total")
                .group("DATE(recorded_at)")
                .order("d")

    # 既存日の結果を { '1' => 24, ... } に
    daily_map = raw_daily.index_by { |r| r.d.day.to_s }

    # 1..末日 を 0 埋めで並べる
    days = (month_start.to_date..month_start.end_of_month).map(&:day).map(&:to_s)
    daily_series = days.map { |d| { day: d, total: (daily_map[d]&.total || 0).to_i } }

    # === 今日のイベント ===
    today_start = Time.zone.today.beginning_of_day
    today_end   = Time.zone.today.end_of_day
    today_rel   = BowlSnapshot.where(device_id: device_id, recorded_at: today_start..today_end)

    today_events = today_rel
                   .order(:recorded_at)
                   .pluck(:recorded_at, :weight_g)
                   .map { |t, g| { time: t.strftime("%H:%M"), g: g.to_f } }

    today_total = today_rel.sum(:weight_g).to_f

    # 過去最大3ヶ月分を対象に1日あたりの平均摂取量(g/日)を算出
    average_daily_intake_last_3_months =
      calculate_average_daily_intake_last_3_months(
        device_id: device_id,
        month_start: month_start,
        month_end: month_end
      )

    render json: {
      todayEvents: today_events,
      dailySeries: daily_series,
      todayTotal:  today_total,
      bowlRemaining: bowl_remaining,
      averageDailyIntakeLast3Months: average_daily_intake_last_3_months
    }
  end

  private


  # 過去最大3ヶ月分を対象に1日あたりの平均摂取量(g/日)を返す
  def calculate_average_daily_intake_last_3_months(device_id:, month_start:, month_end:)
    # 対象期間: 指定月から過去３ヶ月前まで
    period_start = (month_start - 3.months).beginning_of_month
    period_end   = (month_start - 1.month).end_of_month

    # BowlSnapshot の recorded_at をもとに期間内のデータを絞り込む
    scope = BowlSnapshot.where(device_id: device_id, recorded_at: period_start..period_end)
    return 0.0 if scope.blank?

    # DATE(recorded_at) ごとに weight_g を合計し、日別の総量を求める
    daily_totals = scope.group("DATE(recorded_at)").sum(:weight_g)

    return 0.0 if daily_totals.empty?

    total_grams     = daily_totals.values.sum          # 期間内の総摂取量（g）
    days_with_meals = daily_totals.keys.size           # 摂取があった日数

    # 過去最大3ヶ月分を対象にした1日あたりの平均摂取量（g/日）
    (total_grams.to_f / days_with_meals).round(1)
  end
end

File: api/app/Http/Controllers/Api/V1/DashboardController.php
<?php

namespace App\Http\Controllers\Api\V1;

use App\Http\Controllers\Controller;
use Carbon\CarbonImmutable;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;

class DashboardController extends Controller
{
    public function show(Request $request): JsonResponse
    {
        $month = $request->input('month');
        if (!is_string($month) || trim($month) === '') {
            $month = CarbonImmutable::now('Asia/Tokyo')->format('Y-m');
        }

        $monthStart = CarbonImmutable::parse($month . '-01', 'Asia/Tokyo')->startOfMonth();
        $monthEnd = $monthStart->endOfMonth()->endOfDay();

        $deviceId = $request->input('device_id');
        if (!is_string($deviceId) || trim($deviceId) === '') {
            $deviceId = DB::table('devices')->orderBy('id')->value('id');
        }

        if ($deviceId === null) {
            return response()->json([
                'error' => 'device not found',
            ], 404);
        }

        $monthStartUtc = $monthStart->setTimezone('UTC');
        $monthEndUtc = $monthEnd->setTimezone('UTC');

        $lastSnapshot = DB::table('bowl_snapshots')
            ->where('device_id', $deviceId)
            ->orderByDesc('recorded_at')
            ->first();

        $bowlRemaining = $lastSnapshot ? (float) $lastSnapshot->weight_g : 0.0;

        $rawDaily = DB::table('bowl_snapshots')
            ->selectRaw("DATE(timezone('Asia/Tokyo', recorded_at)) AS d, SUM(weight_g) AS total")
            ->where('device_id', $deviceId)
            ->whereBetween('recorded_at', [
                $monthStartUtc->format('Y-m-d H:i:s.u'),
                $monthEndUtc->format('Y-m-d H:i:s.u'),
            ])
            ->groupByRaw("DATE(timezone('Asia/Tokyo', recorded_at))")
            ->orderBy('d')
            ->get();

        $dailyMap = [];
        foreach ($rawDaily as $row) {
            $day = (string) CarbonImmutable::parse($row->d)->day;
            $dailyMap[$day] = (int) $row->total;
        }

        $daysInMonth = $monthStart->daysInMonth;
        $dailySeries = [];
        for ($day = 1; $day <= $daysInMonth; $day++) {
            $dayKey = (string) $day;
            $dailySeries[] = [
                'day' => $dayKey,
                'total' => (int) ($dailyMap[$dayKey] ?? 0),
            ];
        }

        $todayStart = CarbonImmutable::now('Asia/Tokyo')->startOfDay();
        $todayEnd = $todayStart->endOfDay();
        $todayStartUtc = $todayStart->setTimezone('UTC');
        $todayEndUtc = $todayEnd->setTimezone('UTC');

        $todayEventsRows = DB::table('bowl_snapshots')
            ->where('device_id', $deviceId)
            ->whereBetween('recorded_at', [
                $todayStartUtc->format('Y-m-d H:i:s.u'),
                $todayEndUtc->format('Y-m-d H:i:s.u'),
            ])
            ->orderBy('recorded_at')
            ->get(['recorded_at', 'weight_g']);

        $todayEvents = $todayEventsRows->map(function ($row) {
            $time = CarbonImmutable::parse($row->recorded_at, 'UTC')
                ->setTimezone('Asia/Tokyo')
                ->format('H:i');

            return [
                'time' => $time,
                'g' => (float) $row->weight_g,
            ];
        })->all();

        $todayTotal = (float) DB::table('bowl_snapshots')
            ->where('device_id', $deviceId)
            ->whereBetween('recorded_at', [
                $todayStartUtc->format('Y-m-d H:i:s.u'),
                $todayEndUtc->format('Y-m-d H:i:s.u'),
            ])
            ->sum('weight_g');

        $periodStart = $monthStart->subMonths(3)->startOfMonth();
        $periodEnd = $monthStart->subMonths(1)->endOfMonth();
        $periodStartUtc = $periodStart->setTimezone('UTC');
        $periodEndUtc = $periodEnd->setTimezone('UTC');

        $dailyTotals = DB::table('bowl_snapshots')
            ->selectRaw("DATE(timezone('Asia/Tokyo', recorded_at)) AS d, SUM(weight_g) AS total")
            ->where('device_id', $deviceId)
            ->whereBetween('recorded_at', [
                $periodStartUtc->format('Y-m-d H:i:s.u'),
                $periodEndUtc->format('Y-m-d H:i:s.u'),
            ])
            ->groupByRaw("DATE(timezone('Asia/Tokyo', recorded_at))")
            ->get();

        if ($dailyTotals->isEmpty()) {
            $averageDailyIntakeLast3Months = 0.0;
        } else {
            $totalGrams = 0.0;
            foreach ($dailyTotals as $row) {
                $totalGrams += (float) $row->total;
            }

            $daysWithMeals = $dailyTotals->count();
            $averageDailyIntakeLast3Months = $daysWithMeals === 0
                ? 0.0
                : round($totalGrams / $daysWithMeals, 1);
        }

        return response()->json([
            'todayEvents' => $todayEvents,
            'dailySeries' => $dailySeries,
            'todayTotal' => $todayTotal,
            'bowlRemaining' => $bowlRemaining,
            'averageDailyIntakeLast3Months' => $averageDailyIntakeLast3Months,
        ], 200, [], JSON_PRESERVE_ZERO_FRACTION);
    }
}
